### PR TITLE
Avoid a StackOverflow and properly fail on cyclic references in CRD generation

### PR DESCRIPTION
**Pull request by andreaTP**
## Description
This is an initial improvement for #3651 , it mainly adds the failing test and output some human friendly information.
I'm still eager to discuss a proper solution for the underlying issue.

## Type of change
 - [X] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [X] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractJsonSchema.java [hunk 1]
```diff
@@ -361,12 +361,16 @@ private String extractUpdatedNameFromJacksonPropertyIfPresent(Property property)
    * @return the structural schema associated with the specified property
    */
   public T internalFrom(String name, TypeRef typeRef) {
+    return internalFromImpl(name, typeRef, new LinkedList<String>());
+  }
+
+  private T internalFromImpl(String name, TypeRef typeRef, List<String> visited) {
```
Comment (metacosm) [inline crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractJsonSchema.java:367]: If the insertion order is not relevant (and I don't think it is in this instance), it's probably better to use a `Set` instead of a `List`. This will also simplify the lookup part which would end up being simply a call to `contains`…

**Review by metacosm — APPROVED** [review]
This is a good first step but I think that the issue should be addressed at the `sundrio` level…

File: crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractJsonSchema.java (hunk 1 above)
Comment (andreaTP) [reply crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractJsonSchema.java]: Good suggestion, switched to a `Set` for the signature and an `HashSet` for the implementation.

**Comment by andreaTP** [comment]
@metacosm I'm not sure that the underlying issue is in `sundrio`, Kubernetes API and OpenAPI v3 are not allowing cyclic references but it's extremely common to have those in POJOs.
I think we need to come up with a design decision on how to move forward in this case.

File: kubernetes-tests/src/test/java/io/fabric8/crd/generator/zookeeper/v1alpha1/Zookeeper.java [hunk 2]
```diff
@@ -24,7 +24,6 @@
 @Version(value="v1alpha1", storage=false)
 @Group("io.zookeeper")
 public class Zookeeper extends CustomResource<ZookeeperSpec, ZookeeperStatus> implements Namespaced {
-  private ZookeeperSpec spec;
```
Comment (andreaTP) [inline kubernetes-tests/src/test/java/io/fabric8/crd/generator/zookeeper/v1alpha1/Zookeeper.java:27]: is this a pattern that you need / want to support?
In the e2e tests this was throwing the cycle exception (e.g. the fields are parsed twice if they are explicitly declared in this way).

**Comment by andreaTP** [comment]
Moving this PR to Draft since I found more edge cases.

File: crd-generator/api/src/test/java/io/fabric8/crd/example/nocyclic/NoCyclicStatus.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.crd.example.nocyclic;
+
+public class NoCyclicStatus {
+  private String message;
+  // private Ref ref1;
```
Comment (andreaTP) [inline crd-generator/api/src/test/java/io/fabric8/crd/example/nocyclic/NoCyclicStatus.java:20]: This doesn't pass yet, and I'm not sure how to implement it without breaking the signature of public methods, any hint? Or any suggestion?
@iocanel maybe?

Comment (andreaTP) [reply crd-generator/api/src/test/java/io/fabric8/crd/example/nocyclic/NoCyclicStatus.java]: fixed

**Comment by andreaTP** [comment]
@metacosm this should be ready for a final review now.
Tested also locally against the [Keycloak model](https://github.com/keycloak/keycloak/tree/main/core/src/main/java/org/keycloak/representations/idm) and added to the local tests all the edge cases I found, now is good to go 🙂

**Comment by andreaTP** [comment]
cc. @manusa @rohanKanojia this should e ready to merge AFAIU.

**Comment by manusa** [comment]
From the code, I understand that the fix consists on detecting cyclic references and throwing a human readable exception with the identification of the problematic field/entity in case on  is found.

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Review by rohanKanojia — APPROVED** [review]
